tf_encrypted/protocol/aby3/aby3.py

```diff
@@ -0,0 +1,3355 @@
+
+
+def _add_public_private(prot, x, y):
+    assert isinstance(x, ABY3PublicTensor), type(x)
+    assert isinstance(y, ABY3PrivateTensor), type(y)
+    assert x.is_scaled == y.is_scaled, ("Cannot mix different encodings: "
+                                        "{} {}").format(x.is_scaled, y.is_scaled)
+
+    x_on_0, _, x_on_2 = x.unwrapped
+    shares = y.unwrapped
+
+    z = [[None] * 2 for _ in range(3)]
+    with tf.name_scope("add"):
+
+        with tf.device(prot.servers[0].device_name):
+            z[0][0] = shares[0][0] + x_on_0
+            z[0][1] = shares[0][1]
+
+        with tf.device(prot.servers[1].device_name):
+            z[1][0] = shares[1][0]
+            z[1][1] = shares[1][1]
+
+        with tf.device(prot.servers[2].device_name):
+            z[2][0] = shares[2][0]
+            z[2][1] = shares[2][1] + x_on_2
+
+    return ABY3PrivateTensor(prot, z, x.is_scaled, y.share_type)
+
+
+def _add_public_public(prot, x, y):
+    assert isinstance(x, ABY3PublicTensor), type(x)
+    assert isinstance(y, ABY3PublicTensor), type(y)
+    assert x.is_scaled == y.is_scaled, "Cannot add tensors with different scales"
+
+    x_shares = x.unwrapped
+    y_shares = y.unwrapped
+
+    z = [None] * 3
+    with tf.name_scope("add"):
+        for i in range(3):
+            z[i] = x_shares[i] + y_shares[i]
+
+    return ABY3PublicTensor(prot, z[0], z[1], z[2], x.is_scaled, x.share_type)
+
+#
+# sub helpers
+#
+
+
+def _sub_private_private(prot, x, y):
+    assert isinstance(x, ABY3PrivateTensor), type(x)
+    assert isinstance(y, ABY3PrivateTensor), type(y)
+    assert x.is_scaled == y.is_scaled
+
+    z = [[None] * 2 for _ in range(3)]
+    with tf.name_scope("sub"):
+        x_shares = x.unwrapped
+        y_shares = y.unwrapped
+        for i in range(3):
+            with tf.device(prot.servers[i].device_name):
+                z[i][0] = x_shares[i][0] - y_shares[i][0]
+                z[i][1] = x_shares[i][1] - y_shares[i][1]
+
+    return ABY3PrivateTensor(prot, z, x.is_scaled, x.share_type)
+
+
+def _sub_private_public(prot, x, y):
+    assert isinstance(x, ABY3PrivateTensor), type(x)
+    assert isinstance(y, ABY3PublicTensor), type(y)
+    assert x.is_scaled == y.is_scaled
+
+    shares = x.unwrapped
+    y_on_0, _, y_on_2 = y.unwrapped
+
+    z = [[None] * 2 for _ in range(3)]
+    with tf.name_scope("sub"):
+
+        with tf.device(prot.servers[0].device_name):
+            z[0][0] = shares[0][0] - y_on_0
+            z[0][1] = shares[0][1]
+
+        with tf.device(prot.servers[1].device_name):
+            z[1][0] = shares[1][0]
+            z[1][1] = shares[1][1]
+
+        with tf.device(prot.servers[2].device_name):
+            z[2][0] = shares[2][0]
+            z[2][1] = shares[2][1] - y_on_2
+
+    return ABY3PrivateTensor(prot, z, x.is_scaled, x.share_type)
+
+
+def _sub_public_private(prot, x, y):
+    assert isinstance(x, ABY3PublicTensor), type(x)
+    assert isinstance(y, ABY3PrivateTens
```

Comment:
It's in Figure 1 "Single round share truncation protocol trunc2". But as I commented above, maybe it's best to remove this in the current version because it is indeed confusing.

Reply:
I suggest we keep it in for now, suspect that it could be useful soon! :)